### PR TITLE
[Azure.ResourceDetectors] Add appservice service.instance.id

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Fixes #.

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs [hunk 1]
```diff
@@ -29,30 +29,35 @@ public sealed class AppServiceResourceDetector : IResourceDetector
...
+            else
+            {
+                return Resource.Empty;
+            }
         }
-        else
+        catch
         {
-            return Resource.Empty;
+            // TODO: log exception.
```
Comment (cijothomas) [inline src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:60]: why no return here?

File: test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs [hunk 2]
```diff
@@ -30,9 +30,11 @@ public void AppServiceResourceDetectorReturnsResourceWithAttributes()
         try
         {
             Environment.SetEnvironmentVariable("WEBSITE_SITE_NAME", "AzureAppService");
+            Environment.SetEnvironmentVariable("WEBSITE_INSTANCE_ID", "AzureInstance");
```
Comment (cijothomas) [inline test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs:33]: clear this in finally

**Review by rajkumar-rangaraj — APPROVED** [review]
LGTM.

Setting servicename and serviceInstanceId may not be enough for ingestion service to identify it is a Azure App Service. We need to follow up and identify what special attribute constant we need to set in resource to identify it is coming from App Service.

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs [hunk 3]
```diff
@@ -29,29 +29,35 @@ public sealed class AppServiceResourceDetector : IResourceDetector
...
-        List<KeyValuePair<string, object>>? attributeList = null;
+            serviceInstanceId = Environment.GetEnvironmentVariable("WEBSITE_INSTANCE_ID");
+            if (serviceName != null)
+            {
+                attributeList = new();
 
-        if (serviceName != null)
-        {
-            attributeList = new();
+                attributeList.Add(new KeyValuePair<string, object>(ResourceSemanticConventions.AttributeServiceName, serviceName));
```
[the remark this answers is not in the document]
Comment (vishweshbankwar) [reply src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:46]: Depends on the sequence in which they are added. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#merge. The last one takes precedence.

In 1) Detector takes precedence, In 2) myAttributes
1)
```csharp
ResourceBuilder.CreateDefault().AddAttributes(myAttributes).AddDetector(new AppServiceResourceDetector());
```

2) ```csharp
ResourceBuilder.CreateDefault().AddDetector(new AppServiceResourceDetector()).AddAttributes(myAttributes);
```

[the remark this answers is not in the document]
Comment (Kielek) [reply src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:46]: `ResourceBuilder.CreateDefault()` is always detecting information from `OTEL_SERVICE_NAME`. It is.
When you will work on documentation, you should include information that if someone will apply `ResourceBuilder.CreateDefault().AddDetector(new AppServiceResourceDetector())` or just call `AddDetector` on Tracer/Metrics/Logger Builder then the service name will be overridden on azure env.

**Review by Kielek — APPROVED** [review]

**Review by cijothomas — APPROVED** [review]

**Pull request merged**